# Lab: Building a Reproducible Analysis Workflow

In the first part of this lab, you will use official National Park Service data to answer a guided question. In the second part, you will find a dataset and conduct an analysis of your own.

Use the workflow from lecture throughout the lab:

1. **Question**
2. **Data**
3. **Operation**
4. **Check**
5. **Evidence**
6. **Conclusion**
7. **Limitation**


## Learning goals

By the end of this lab, you should be able to:

- organize an investigation with the reproducible analysis workflow;
- inspect a dataset before analyzing it;
- group, summarize, check, and sort observations with Pandas;
- formulate a research question that can be answered with data; and
- distinguish evidence, conclusions, and limitations.


## Using AI tools responsibly

You may use AI to ask questions, debug code, or receive feedback. You remain responsible for understanding your work, verifying suggestions against the data, and citing the source of your dataset. At the end, disclose whether you used an AI tool. Using AI is not required.


# Part 1: Florida park visitation


## Background

The National Park Service (NPS) publishes estimates of recreational visits to its units. These estimates help parks understand patterns in public use and plan staffing, maintenance, transportation, and visitor services.

This course file contains monthly estimates for five Florida NPS units from 2015 through 2025:

- Big Cypress National Preserve
- Biscayne National Park
- Canaveral National Seashore
- Dry Tortugas National Park
- Everglades National Park

A recreational visit is a visit, not necessarily one unique person. A person who visits several times may be counted several times.


| Feature | Description |
| --- | --- |
| `park_code` | Four-letter NPS unit code |
| `park` | Name of the NPS unit |
| `year` | Calendar year of the observation |
| `month_number` | Month represented as an integer from 1 through 12 |
| `month` | Month name |
| `recreation_visits` | Estimated recreational visits during that park-month |


## Step 1: Question

**After the decline in July 2020, when did Canaveral National Seashore's July recreational visitation first return to or exceed its average July visitation from 2015–2019?**

The 2015–2019 mean provides a pre-pandemic comparison point. In this question, “recovery” means the first July from 2020 onward with visitation at or above that baseline. This definition does not require every later July to remain above the baseline.


## Step 2: Data

Import Pandas using the alias `pd`. Read `data/nps_florida_park_visitation_2015_2025.csv` into a DataFrame named `park_visits`, then display its first five rows.


In [1]:
# Write your code here.
import pandas as pd

park_visits = pd.read_csv("data/nps_florida_park_visitation_2015_2025.csv")


- Head of the pandas

In [2]:
park_visits.head()

,park_code,park,year,month_number,month,recreation_visits
0,BICY,Big Cypress National Preserve,2015,1,January,114045
1,BICY,Big Cypress National Preserve,2015,2,February,144112
2,BICY,Big Cypress National Preserve,2015,3,March,153084
3,BICY,Big Cypress National Preserve,2015,4,April,113880
4,BICY,Big Cypress National Preserve,2015,5,May,72510


Display the last five rows.


In [3]:
# Write your code here.
park_visits.tail()


,park_code,park,year,month_number,month,recreation_visits
655,EVER,Everglades National Park,2025,8,August,50464
656,EVER,Everglades National Park,2025,9,September,32290
657,EVER,Everglades National Park,2025,10,October,19008
658,EVER,Everglades National Park,2025,11,November,53416
659,EVER,Everglades National Park,2025,12,December,70926


Display information about the DataFrame and calculate the number of missing observations in each feature.


In [4]:
park_visits["park"].unique()

<ArrowStringArray>
['Big Cypress National Preserve',        'Biscayne National Park',
   'Canaveral National Seashore',    'Dry Tortugas National Park',
      'Everglades National Park']
Length: 5, dtype: str

In [5]:
# Write your code here.
park_visits["year"].unique()

array([2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024, 2025])

In [6]:
# Here are the parks that are calculated
park_visits.groupby(["park"]).count()

,park_code,year,month_number,month,recreation_visits
park,,,,,
Big Cypress National Preserve,132,132,132,132,132
Biscayne National Park,132,132,132,132,132
Canaveral National Seashore,132,132,132,132,132
Dry Tortugas National Park,132,132,132,132,132
Everglades National Park,132,132,132,132,132


In [7]:
# some stats
park_visits["recreation_visits"].describe()

count       660.000000
mean      83304.665152
std       73216.283918
min         176.000000
25%       30373.750000
50%       63529.500000
75%      121226.750000
max      401190.000000
Name: recreation_visits, dtype: float64

In [8]:
park_visits.info()

<class 'pandas.DataFrame'>
RangeIndex: 660 entries, 0 to 659
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   park_code          660 non-null    str  
 1   park               660 non-null    str  
 2   year               660 non-null    int64
 3   month_number       660 non-null    int64
 4   month              660 non-null    str  
 5   recreation_visits  660 non-null    int64
dtypes: int64(3), str(3)
memory usage: 54.1 KB


In [9]:
park_visits.isnull().sum()

park_code            0
park                 0
year                 0
month_number         0
month                0
recreation_visits    0
dtype: int64

**What does each row represent?**

Your answer:
- Each row represents the number of visits to a specified park on any given month.


**Are any observations missing from the six features?**

Your answer:
- No missing observations from the six features


## Step 3: Operation

Create one Boolean condition that identifies Canaveral National Seashore and another that identifies July.


In [10]:
# Write your code here.
parkCanaveral = park_visits["park"] == "Canaveral National Seashore"
monthJuly = park_visits["month_number"] == 7

Combine the two conditions with `&`. Use `.loc` to select the `year` and `recreation_visits` features, make a copy named `canaveral_july`, and display it.


In [11]:
# Write your code here.
canaveral_july = park_visits.loc[parkCanaveral & monthJuly][["year", "recreation_visits"]]

canaveral_july

,year,recreation_visits
270,2015,163169
282,2016,216437
294,2017,138018
306,2018,189285
318,2019,162947
330,2020,112898
342,2021,206636
354,2022,233352
366,2023,173296
378,2024,176242


Create `pre_pandemic_condition`, which is `True` for years from 2015 through 2019. Use it with `.loc` to calculate the mean of `recreation_visits`, and store the result as `pre_pandemic_july_mean`.

Display `pre_pandemic_july_mean`.


In [12]:
# Write your code here.

pre_pandemic_condition = (park_visits["year"] <= 2019)&(park_visits["year"]>=2015) & monthJuly
pre_pandemic_july_mean = park_visits.loc[pre_pandemic_condition]["recreation_visits"].mean()

pre_pandemic_july_mean

# park_visits["pre_pandemic_condition"] = park_visits["year"].apply(lambda x: 2015 <= x <= 2019)

np.float64(72117.88)

Create a condition that identifies rows from 2020 onward whose `recreation_visits` are greater than or equal to `pre_pandemic_july_mean`. Use `.loc` to create `recovery_years`, then display it in ascending year order.


In [13]:
# Write your code here.
after_pandemic = park_visits["year"] > 2019
greaterThanprePandemic = park_visits["recreation_visits"] >= pre_pandemic_july_mean
recovery_years = park_visits.loc[monthJuly &greaterThanprePandemic & after_pandemic & parkCanaveral].sort_values("year",ascending = True)
recovery_years

,park_code,park,year,month_number,month,recreation_visits
330,CANA,Canaveral National Seashore,2020,7,July,112898
342,CANA,Canaveral National Seashore,2021,7,July,206636
354,CANA,Canaveral National Seashore,2022,7,July,233352
366,CANA,Canaveral National Seashore,2023,7,July,173296
378,CANA,Canaveral National Seashore,2024,7,July,176242
390,CANA,Canaveral National Seashore,2025,7,July,253403


## Step 4: Check

Display the number of rows in `canaveral_july`, its smallest year, and its greatest year.


In [14]:
# Write your code here.
print(canaveral_july.count())
print(canaveral_july.min())
print(canaveral_july.max())

year                 11
recreation_visits    11
dtype: int64
year                   2015
recreation_visits    112898
dtype: int64
year                   2025
recreation_visits    253403
dtype: int64


**Do these results confirm that the filtered data contains one July observation for every year from 2015 through 2025?**

Your answer: **YES**


Count the observations used to calculate `pre_pandemic_july_mean`.


In [15]:
# Write your code here.
# pre_pandemic_condition = (park_visits["year"] <= 2019)&(park_visits["month_number"]==7)
park_visits[parkCanaveral].loc[pre_pandemic_condition].count()

park_code            5
park                 5
year                 5
month_number         5
month                5
recreation_visits    5
dtype: int64

**Does the baseline use the expected five years?**

Your answer: Yes


## Step 5: Evidence

Display the first row of `recovery_years`.


In [16]:
# Write your code here.
recovery_years[:1]

,park_code,park,year,month_number,month,recreation_visits
330,CANA,Canaveral National Seashore,2020,7,July,112898


**What was the 2015–2019 mean, and which year first met or exceeded it after the July 2020 decline? Report that year's July visitation.**

Your answer: The mean for 2015-2019 for recreational visits at Canaveral National Seashore was 72117.88. Immediately after the pandemic, recreational visits to the park increased by 56% with a staggering 112,898 visits in July alone.


## Step 6: Conclusion


**Write one sentence that answers the research question using the evidence.**

Your answer: After the pandemic, there was an increase in park visitation in July compared to previous years. 

## Step 7: Limitation


In [17]:
recovery_years["percentageIncrease"] = ((recovery_years["recreation_visits"]-pre_pandemic_july_mean)/pre_pandemic_july_mean)*100
recovery_years

,park_code,park,year,month_number,month,recreation_visits,percentageIncrease
330,CANA,Canaveral National Seashore,2020,7,July,112898,56.546476
342,CANA,Canaveral National Seashore,2021,7,July,206636,186.525339
354,CANA,Canaveral National Seashore,2022,7,July,233352,223.570244
366,CANA,Canaveral National Seashore,2023,7,July,173296,140.295472
378,CANA,Canaveral National Seashore,2024,7,July,176242,144.380450
390,CANA,Canaveral National Seashore,2025,7,July,253403,251.373335


**Does first exceeding the baseline mean that every later July remained above it? Use the data to explain.**

Your answer: Yes, recreational visits have remained above the baseline pre-pandemic.

**Why does this analysis alone not establish that the pandemic caused the 2020 decline?**

Your answer: Because there the amount of visitors continue to grow after 2020, the pandemic cannot be the single cause of any decline in park visits.


# Part 2: Your own investigation


## Find a dataset

Find a CSV dataset that interests you. Kaggle is one possible source, but you may use another reputable source. Choose a dataset that:

- has a clearly identified publisher or creator;
- includes documentation explaining the observations and features;
- contains at least one categorical and one quantitative feature;
- is appropriate to share in a class assignment;
- does not contain private or personally identifying information; and
- is small enough to run comfortably in this notebook.

Save the CSV in the notebook's `data` folder before continuing.


**What is the dataset's title, publisher or creator, and direct source URL?**

Your answer:


**What does one row represent?**

Your answer:


## Step 1: Question

Formulate one research question that can be answered with your dataset. The question must require a calculation or comparison, not merely looking up one value.


**What is your research question?**

Your answer:


## Step 2: Data

Read your CSV into a clearly named DataFrame. Display the first and last five rows, information about the DataFrame, and the missing-value count for every feature.


In [18]:
# Write your code here.


**Which observations and features will you use to answer your question?**

Your answer:


## Step 3: Operation

Write the code needed to prepare, filter, group, summarize, or sort the data. Store the final evidence in a clearly named object and display it.


In [19]:
# Write your code here.


**Explain how your operations connect the selected data to your research question.**

Your answer:


## Step 4: Check

Write at least one check that could reveal an error in your analysis, then display its result.


In [20]:
# Write your code here.


**What did you check, and did the result meet your expectation?**

Your answer:


## Step 5: Evidence


**Which specific value or comparison in your output answers the question?**

Your answer:


## Step 6: Conclusion


**What conclusion is supported by the evidence?**

Your answer:


## Step 7: Limitation


**What should a reader avoid concluding from your analysis, and why?**

Your answer:


## AI-use reflection


**State whether you used an AI tool. If you did, describe one suggestion you checked and what you accepted, changed, or rejected. If you did not, state that you did not use one.**

Your response:


## Submission checklist

Before submitting, confirm that all cells run in order, the independent dataset and its source are included, every answer is complete, numerical claims match the output, conclusions are appropriately limited, and the AI-use reflection is complete.
